File: legacy_py/arbitrage/chains.py
"""Chain normalization helpers."""

from __future__ import annotations

import re
from typing import Dict, Iterable, Tuple

_CHAIN_TOKEN_RE = re.compile(r"[^A-Z0-9]")
_CHAIN_ALIASES: Dict[str, str] = {
    "APTOS": "APT",
    "APT": "APT",
    "ASTAR": "ASTR",
    "ALGORAND": "ALGO",
    "ALLORA": "ALLO",
    "AELF": "ELF",
    "AVAILDA": "AVAIL",
    "ARBITRUMONE": "ARBONE",
    "ARBONE": "ARBONE",
    "ARBITRUMNOVA": "ARBNOVA",
    "ARBNOVA": "ARBNOVA",
    "ARWEAVE": "AR",
    "AVALANCHECCHAIN": "AVAXC",
    "AVALANCHEC": "AVAXC",
    "AVAXC": "AVAXC",
    "BABYLON": "BABY",
    "CELESTIA": "TIA",
    "BASE": "BASE",
    "BERACHAIN": "BERA",
    "BINANCESMARTCHAIN": "BEP20",
    "BNBSMARTCHAIN": "BEP20",
    "BOUNCEBIT": "BB",
    "BSC": "BEP20",
    "BEP20": "BEP20",
    "BITCOIN": "BTC",
    "BITCOINCASH": "BCH",
    "BITCOINSV": "BSV",
    "BITTENSOR": "TAO",
    "BTC": "BTC",
    "CSPR": "CSPR",
    "CASPER": "CSPR",
    "CARDANO": "ADA",
    "CELOL2": "CELO",
    "CFXESPACE": "CFXEVM",
    "CHILIZCHAIN": "CHZ2",
    "COREDAO": "CORE",
    "COSMOS": "ATOM",
    "CRONOS": "CRO",
    "DOGECOIN": "DOGE",
    "DYDXCHAIN": "DYDX",
    "ECASH": "XEC",
    "ECLIPSE": "ES",
    "ENJIN": "ENJ",
    "ERC20": "ERC20",
    "ETH": "ERC20",
    "ETHEREUM": "ERC20",
    "ETHEREUMCLASSIC": "ETC",
    "ETHEREUMPOW": "ETHW",
    "FILECOIN": "FIL",
    "FLARE": "FLR",
    "G": "G",
    "GRAVITY": "G",
    "HIPPOPROTOCOL": "HP",
    "ICON": "ICX",
    "INTERNETCOMPUTER": "ICP",
    "INJECTIVE": "INJ",
    "INITIA": "INIT",
    "IOTEX": "IOTX",
    "KAIA": "KAIA",
    "KLAYTN": "KAIA",
    "MANTRA": "OM",
    "LUNA": "LUNA",
    "TERRA": "LUNA",
    "MATIC": "MATIC",
    "MEDIBLOC": "MED",
    "METALL2": "MTLETH",
    "MONAD": "MON",
    "MERLINCHAIN": "MERLBTC",
    "MULTIVERSEX": "EGLD",
    "NEARPROTOCOL": "NEAR",
    "NEON3": "GAS",
    "NERVOSNETWORK": "CKB",
    "NILLION": "NIL",
    "OASYS": "OAS",
    "OPTIMISM": "OP",
    "OP": "OP",
    "ONTOLOGY": "ONG",
    "OSMOSIS": "OSMO",
    "PLASMA": "XPL",
    "POCKETNETWORK": "POKT",
    "POLYGON": "MATIC",
    "POLYMESH": "POLYX",
    "PROTON": "XPR",
    "QUARKCHAIN": "QKC",
    "RAVEN": "RVN",
    "REINETWORK": "REI",
    "RON": "RON",
    "RONIN": "RON",
    "SIACOIN": "SC",
    "SEINETWORK": "SEI",
    "SHENTU": "CTK",
    "S": "S",
    "SONIC": "S",
    "SONGBIRD": "SGB",
    "SOLAR": "SXP",
    "SOL": "SOL",
    "SOLANA": "SOL",
    "SOMNIA": "SOMI",
    "SOPHON": "SOPH",
    "STELLAR": "XLM",
    "STORY": "IP",
    "STRATIS": "STRAX",
    "STRKETH": "STRKETH",
    "STARKNET": "STRKETH",
    "STACKS": "STX",
    "SWELLCHAIN": "SWELL",
    "TRC20": "TRC20",
    "TRON": "TRC20",
    "TRX": "TRC20",
    "TEZOS": "XTZ",
    "THUNDERCORE": "TT",
    "THETA": "THETA",
    "THETANETWORK": "THETA",
    "VAULTA": "A",
    "VECHAIN": "VET",
    "WAX": "WAXP",
    "WORLDCHAIN": "WLD",
    "ZETA": "ZETA",
    "ZETACHAIN": "ZETA",
    "ZKSERA": "ZKSERA",
    "ZKSYNCERA": "ZKSERA",
    "ZILLIQA": "ZIL",
    "ZIRCUIT": "ZRC",
}


def normalize_chain_name(name: str) -> str:
    token = _CHAIN_TOKEN_RE.sub("", (name or "").upper())
    return _CHAIN_ALIASES.get(token, token)


def common_chains(a: Iterable[str], b: Iterable[str]) -> Tuple[str, ...]:
    """Best-effort intersection of chain names."""
    a_map = _canonical_chain_map(a)
    b_map = _canonical_chain_map(b)
    keys = sorted(set(a_map) & set(b_map))
    return tuple(a_map[k] for k in keys)


def common_chain_pairs(a: Iterable[str], b: Iterable[str]) -> Tuple[Tuple[str, str], ...]:
    """Best-effort intersection returning `(a_name, b_name)` pairs."""
    a_map = _canonical_chain_map(a)
    b_map = _canonical_chain_map(b)
    keys = sorted(set(a_map) & set(b_map))
    return tuple((a_map[k], b_map[k]) for k in keys)


def _canonical_chain_map(names: Iterable[str]) -> Dict[str, str]:
    out: Dict[str, str] = {}
    for name in names:
        if not name:
            continue
        key = normalize_chain_name(name)
        if key and key not in out:
            out[key] = name
    return out

File: src/transfer/chains.ts
/**
 * @fileoverview Blockchain chain name normalization and matching utilities.
 * Provides consistent chain naming across different exchanges.
 */

/**
 * Chain name aliases mapping various exchange-specific names to canonical names.
 * Keys are uppercase with spaces removed, values are canonical chain names.
 */
const CHAIN_ALIASES: Readonly<Record<string, string>> = {
  // Ethereum
  ERC20: "ETH",
  ETHEREUM: "ETH",
  ETH: "ETH",
  // Arbitrum
  ARBITRUM: "ARBITRUM",
  ARBITRUMONE: "ARBITRUM",
  ARB: "ARBITRUM",
  ARBONE: "ARBITRUM",
  // Optimism
  OPTIMISM: "OPTIMISM",
  OP: "OPTIMISM",
  // BSC
  BSC: "BSC",
  BEP20: "BSC",
  BNBSMARTCHAIN: "BSC",
  "BNBSMARTCHAIN(BEP20)": "BSC",
  BINANCE: "BSC",
  // Tron
  TRON: "TRON",
  TRC20: "TRON",
  TRX: "TRON",
  // Solana
  SOL: "SOL",
  SOLANA: "SOL",
  // Aptos
  APT: "APTOS",
  APTOS: "APTOS",
  // Polygon
  MATIC: "POLYGON",
  POLYGON: "POLYGON",
  POLYGONPOS: "POLYGON",
  // TON
  TON: "TON",
  TONCOIN: "TON",
  // Bitcoin
  BTC: "BTC",
  BITCOIN: "BTC",
  // Bitcoin Cash
  BCH: "BCH",
  BITCOINCASH: "BCH",
  // ICON
  ICX: "ICX",
  ICON: "ICX",
  // aelf
  ELF: "ELF",
  AELF: "ELF",
  // WAX
  WAX: "WAX",
  WAXP: "WAX",
  // Theta
  THETA: "THETA",
  THETANETWORK: "THETA",
  // Waves
  WAVES: "WAVES",
  // Ontology
  ONT: "ONT",
  ONTOLOGY: "ONT",
  // Qtum
  QTUM: "QTUM",
  // Ravencoin
  RVN: "RVN",
  RAVENCOIN: "RVN",
  // Decred
  DCR: "DCR",
  DECRED: "DCR",
  // Zcash
  ZEC: "ZEC",
  ZCASH: "ZEC",
  // Horizen
  ZEN: "ZEN",
  HORIZEN: "ZEN",
  // Stacks
  STX: "STX",
  STACKS: "STX",
  // Flux
  FLUX: "FLUX",
  // Nervos
  CKB: "CKB",
  NERVOS: "CKB",
  NERVOSNETWORK: "CKB",
  // Arweave
  AR: "AR",
  ARWEAVE: "AR",
  // Oasis
  ROSE: "ROSE",
  OASIS: "ROSE",
  OASISNETWORK: "ROSE",
  // Astar
  ASTR: "ASTAR",
  ASTAR: "ASTAR",
  // MultiversX (Elrond)
  EGLD: "EGLD",
  MULTIVERSX: "EGLD",
  ELROND: "EGLD",
  // Ethereum Classic
  ETC: "ETC",
  ETHEREUMCLASSIC: "ETC",
  // Neo
  NEO: "NEO",
  NEON3: "NEO",
  "NEON3(N3)": "NEO",
  GAS: "GAS",
  NEOGAS: "GAS",
  // Zilliqa
  ZIL: "ZIL",
  ZILLIQA: "ZIL",
  // zkSync
  ZKSYNC: "ZKSYNC",
  ZKSERA: "ZKSYNC",
  ZKSYNCERA: "ZKSYNC",
  // Chiliz
  CHZ: "CHZ",
  CHZ2: "CHZ",
  CHILIZ: "CHZ",
  CHILIZCHAIN: "CHZ",
  // Bittensor
  TAO: "TAO",
  BITTENSOR: "TAO",
  // Story (IP)
  IP: "IP",
  STORY: "IP",
  // Monad
  MON: "MON",
  MONAD: "MON",
  // Cosmos ecosystem
  ATOM: "ATOM",
  COSMOS: "ATOM",
  // Avalanche
  AVAX: "AVAX",
  AVALANCHE: "AVAX",
  AVAXC: "AVAX",
  "AVALANCHEC-CHAIN": "AVAX",
  // Sui
  SUI: "SUI",
  // Base
  BASE: "BASE",
  // Linea
  LINEA: "LINEA",
  // Scroll
  SCROLL: "SCROLL",
  // Blast
  BLAST: "BLAST",
  // SEI
  SEI: "SEI",
  // Celestia
  TIA: "TIA",
  CELESTIA: "TIA",
  // NEAR
  NEAR: "NEAR",
  // Hedera
  HBAR: "HBAR",
  HEDERA: "HBAR",
  // Cardano
  ADA: "ADA",
  CARDANO: "ADA",
  // XRP
  XRP: "XRP",
  RIPPLE: "XRP",
  // Polkadot
  DOT: "DOT",
  POLKADOT: "DOT",
  // Kaspa
  KAS: "KAS",
  KASPA: "KAS",
  // DOGE
  DOGE: "DOGE",
  DOGECOIN: "DOGE",
  // Litecoin
  LTC: "LTC",
  LITECOIN: "LTC",
  // Initia
  INIT: "INIT",
  INITIA: "INIT",
  // Merlin
  MERL: "MERL",
  MERLINCHAIN: "MERL",
  MERLBTC: "MERL",
  // Shentu (CTK)
  CTK: "CTK",
  SHENTU: "CTK",
  // Sophon
  SOPH: "SOPH",
  SOPHON: "SOPH",
  // Starknet
  STRK: "STRK",
  STARKNET: "STRK",
  // Mantle
  MNT: "MNT",
  MANTLE: "MNT",
  // Kaia (previously Klaytn)
  KAIA: "KAIA",
  KLAYTN: "KAIA",
  KLAY: "KAIA",
  // Mode
  MODE: "MODE",
  // Zora
  ZORA: "ZORA",
  // Celo
  CELO: "CELO",
  // Flow
  FLOW: "FLOW",
  // Fantom
  FTM: "FTM",
  FANTOM: "FTM",
  // Algorand
  ALGO: "ALGO",
  ALGORAND: "ALGO",
  // Cronos
  CRO: "CRO",
  CRONOS: "CRO",
  // ICP
  ICP: "ICP",
  INTERNETCOMPUTER: "ICP",
  // VeChain
  VET: "VET",
  VECHAIN: "VET",
  // IOTA
  IOTA: "IOTA",
  // Filecoin
  FIL: "FIL",
  FILECOIN: "FIL",
  // Bitcoin SV
  BSV: "BSV",
  BITCOINSV: "BSV",
  // dYdX
  DYDX: "DYDX",
  DYDXCHAIN: "DYDX",
  // Injective
  INJ: "INJ",
  INJECTIVE: "INJ",
  // Osmosis
  OSMO: "OSMO",
  OSMOSIS: "OSMO",
  // Akash
  AKT: "AKT",
  AKASH: "AKT",
  // Secret
  SCRT: "SCRT",
  SECRET: "SCRT",
  // Kujira
  KUJI: "KUJI",
  KUJIRA: "KUJI",
  // Juno
  JUNO: "JUNO",
  // Terra
  LUNA: "LUNA",
  TERRA: "LUNA",
  // Sommelier
  SOMM: "SOMM",
  SOMMELIER: "SOMM",
  // Vaulta (A)
  A: "VAULTA",
  VAULTA: "VAULTA",
  // Enjin
  ENJ: "ENJ",
  ENJIN: "ENJ",
  // Stellar
  XLM: "XLM",
  STELLAR: "XLM",
  // Tezos
  XTZ: "XTZ",
  TEZOS: "XTZ",
  XTZEVM: "XTZ",
  // eCash
  XEC: "XEC",
  ECASH: "XEC",
  // MediBloc
  MED: "MED",
  MEDIBLOC: "MED",
  // Proton
  XPR: "XPR",
  PROTON: "XPR",
  // Aergo
  AERGO: "AERGO",
  // Solar
  SXP: "SXP",
  SOLAR: "SXP",
  // IOST
  IOST: "IOST",
  // Bifrost
  BFC: "BFC",
  BIFROST: "BFC",
  // Metal
  MTL: "MTL",
  METALL2: "MTL",
  MTLETH: "MTL",
};

/**
 * Normalizes a chain name to its canonical form.
 *
 * @param name - The chain name to normalize (case-insensitive)
 * @returns The canonical chain name, or the input uppercased if no alias found
 *
 * @example
 * ```typescript
 * normalizeChainName("ERC20");     // Returns "ETH"
 * normalizeChainName("trc20");     // Returns "TRON"
 * normalizeChainName("Ethereum");  // Returns "ETH"
 * normalizeChainName("Unknown");   // Returns "UNKNOWN"
 * ```
 */
export function normalizeChainName(name: string): string {
  const raw = (name || "").trim().toUpperCase();
  if (!raw) {
    return "";
  }
  const compact = raw.replace(/\s+/g, "");
  return CHAIN_ALIASES[compact] ?? compact;
}

/**
 * Finds common chains between two sets of chain names.
 *
 * @param a - First set of chain names
 * @param b - Second set of chain names
 * @returns Array of normalized chain names that appear in both sets
 *
 * @example
 * ```typescript
 * commonChains(["ERC20", "TRC20"], ["ETH", "TRON"]);
 * // Returns ["ETH", "TRON"]
 * ```
 */
export function commonChains(a: Iterable<string>, b: Iterable<string>): string[] {
  const normalizedA = new Set(
    Array.from(a, (x) => normalizeChainName(x)).filter(Boolean)
  );
  const normalizedB = new Set(
    Array.from(b, (x) => normalizeChainName(x)).filter(Boolean)
  );
  return Array.from(normalizedA)
    .filter((x) => normalizedB.has(x))
    .sort();
}

/**
 * Finds common chains between two sets and returns original name pairs.
 *
 * @param a - First set of chain names
 * @param b - Second set of chain names
 * @returns Array of [originalA, originalB] pairs for matching chains
 *
 * @example
 * ```typescript
 * commonChainPairs(["ERC20", "TRC20"], ["Ethereum", "TRON"]);
 * // Returns [["ERC20", "Ethereum"], ["TRC20", "TRON"]]
 * ```
 */
export function commonChainPairs(
  a: Iterable<string>,
  b: Iterable<string>
): Array<[string, string]> {
  const normalizedToOriginalA = new Map<string, string>();
  for (const x of a) {
    const normalized = normalizeChainName(x);
    if (normalized) {
      normalizedToOriginalA.set(normalized, x);
    }
  }

  const result: Array<[string, string]> = [];
  for (const y of b) {
    const normalized = normalizeChainName(y);
    const originalA = normalizedToOriginalA.get(normalized);
    if (originalA && normalized) {
      result.push([originalA, y]);
    }
  }
  return result;
}

/**
 * Checks if a chain name is a known blockchain.
 *
 * @param name - The chain name to check
 * @returns True if the chain is recognized
 */
export function isKnownChain(name: string): boolean {
  const normalized = normalizeChainName(name);
  return normalized !== "" && Object.values(CHAIN_ALIASES).includes(normalized);
}
